### ISSUE TITLE
Calls client.quit() on a redis connection that it doesn't own

### ISSUE THREAD

**Issue by webhype** [wankdanker/node-redis-index#4]
Thanks for a great library, it doesn't seem to get the traction here that it deserves in my opinion. So much work went into this!! Perhaps people are deterred by lack of documentation? (hint hint hahah)

About this issue: In my application, I override createClient like so, as suggested in the documentation:

				RedisIndex.createClient = function () {
					return myStuff.redis;
				};

The expected behavior is that the `RedisIndex` class wouldn't assume that it owns the connection. In fact, I am trying to connection-pool, meaning I am trying to avoid opening more and more redundant sockets to Redis, or whatever other services I am using.

However, your lib calls `client.quit()` via `index.end()`quite a bit, even under non-error conditions, example  `redis-index.js`line 150:

		client.multi(commands).exec(function (err, result) {
			end(err, result);
		});

`err` is generally null, but end() closes the connection regardless, even though I still need it. Even on error, quitting "my" Redis connection is kinda rude, when `RedisIndex` can't be sure it opened it from scratch. 🤣 Perhaps leave that to the caller? Or add a hook to close connection, if and only if error, so the caller can manage it with some kind of global use count?

For some reason I am only able to replicate this bug when running using await / Promises. I assume without that, due to a race condition, "most of the time", the end() calls don't happen before my connection is closed from underneath me.

Here is some example code that fails "most of the time", at random moments during execution, unless I comment out `client.quit()` in lines 290 and 362:

	const RedisIndex = require('lib/redis-index');

	const ri = RedisIndex.createIndex({
		key: 'ri:test',	// Redis key namespace
		index: 'unique',	// the attribute of the document whose value uniquely identifies that object
		schema: {
			name: {
				literal: true
			},
			boolz: {type: 'boolean'}
		},
		schemaOnly: true, // only index fields that exist in the schema
		storeObject: false,	// store the document in redis or not
		createClient: () => { return myStuff.redis; },
		fullText: null
	});

	const docs = [
		{unique: 1, name: 'steve', description: 'hello there thing', type: 1, when: new Date(), value: 100, boolz: false},
		{unique: 2, name: 'sam', description: 'this thing is named sam', type: 1, when: new Date(), value: 50, boolz: false},
		{unique: 3, name: 'george', description: 'and here is one george', type: 2, when: new Date(), value: 75, boolz: true},
		{unique: 4, name: 'brenda mcgill', description: 'i am a hungry thing', type: 2, when: new Date(), value: 30, boolz: false},
	];
	let proms = [];

	for (let d of docs) {
		proms.push(ri.add(d));
	}
	await Promise.all(proms);

	/** @type {RedisIndexSearch} */
	let ris = await ri.search({name: 'sam'});
	ris = await ris.or({unique: 1});
	const result = await ris.exec();
	console.log(result);

	// Remove all 

	for (let d of docs) {
		proms.push(ri.del(d));
	}
	await Promise.all(proms);

Hope this helps!

**Comment by wankdanker** [comment]
Thank you for pointing this out. I have one use case for this module and I have always just wanted it to close the connections when done. But, what you say makes complete sense and is how I would prefer it to work, honestly.

Rather than making `RedisIndex` and `RedisIndexSearch` event emitters that we would need to subscribe to for end/close events, what would you think about something similar to this:

```js
//default implementation to be included with the module
RedisIndex.closeClient = function (client) {
   client.quit();
};
```

```js
//you could over ride it with this
RedisIndex.closeClient = function (client) {
   //do nothing, or maybe pop this client back in to the pool as an unused connectin
};
```

**Comment by webhype** [comment]
I concur. I guess if you want to be pedantic, it should be called `destroyClient()` because `destroy` is the opposite of `create` in computer speak 😂. The super correct way would be an `obtainClient()` / `releaseClient()` pair, or something like that, but that would no longer be backwards compatible.

It would be helpful to pass `err` to such a destroyClient() hook so, as a caller, I can decide if there was a serious enough reason to close my Redis connection immediately, in my hook, on my own terms, or if it was just a successful task completion, in which case I would just like to recycle the same connection, i.e., I would not close it in that case.

I hope that makes sense... thanks again for your support!

**Comment by wankdanker** [comment]
Totally agree with you. Will add `err` to the hook. I'm fine with renaming them to `obtainClient()` / `releaseClient()`. I can do it backwards compatible.

This will be released with a major version bump though, because the update to node-redis@3.0.2 breaks backwards compatibility with node < 6.

**Comment by wankdanker** [comment]
This update has been published to npm as redis-index@4.0.0
